### PR TITLE
feat: canister detail modals with custom event

### PR DESCRIPTION
**Pull request by peterpeterparker**
# Motivation

Replace usage of context store to trigger the canister details modal opening, closing and passing extra data.

# Changes

- use `CustomEvent` to trigger opening of modals used in "Canister" detail page
- adapt `Intersection` types to pattern use in this PR
- remove few unused imports

### CODE REVIEW

File: frontend/src/lib/types/canister-detail.modal.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  | "detach"
+  | "add-controller"
+  | "remove-controller";
+
+export interface CanisterDetailsModal<D = void> {
+  type: CanisterDetailsModalType;
+  detail?: D;
+}
+
+export type CanisterDetailsModalRemoveController = Required<
```
Comment (lmuntaner) [inline frontend/src/lib/types/canister-detail.modal.ts:14]: Is this the same as?

```
export type CanisterDetailsModalRemoveController = CanisterDetailsModal<{ controller: string }> & {
  detail: { controller: string };
};
```

If we ever add properties to CanisterDetailsModal we would need to pick them as well.

Also, you used an interface that you never extend, just pick from it. It looks weird to me. Is there an advantage over extending it?

File: frontend/src/tests/lib/components/canister-detail/RemoveCanisterControllerButton.spec.ts [hunk 2]
```diff
@@ -87,32 +87,4 @@ describe("RemoveCanisterControllerButton", () => {
     expect(removeController).toBeCalled();
     expect(reloadDetailsMock).toBeCalled();
   });
-
-  it("clear selected controller after remove", async () => {
```
Comment (lmuntaner) [inline frontend/src/tests/lib/components/canister-detail/RemoveCanisterControllerButton.spec.ts:91]: Why did you remove this test?

File: frontend/src/lib/types/canister-detail.modal.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+export type CanisterDetailsModalType =
+  | "add-cycles"
+  | "detach"
+  | "add-controller"
+  | "remove-controller";
+
+export interface CanisterDetailsModal<D = void> {
+  type: CanisterDetailsModalType;
+  detail?: D;
```
Comment (lmuntaner) [inline frontend/src/lib/types/canister-detail.modal.ts:11]: The term `detail` is also used in the `emit`. Shall we call this `data` or `payload`?

**Review by lmuntaner — COMMENTED** [review]
Thanks! A few comments

File: frontend/src/tests/lib/components/canister-detail/RemoveCanisterControllerButton.spec.ts (hunk 2 above)
Comment (peterpeterparker) [reply frontend/src/tests/lib/components/canister-detail/RemoveCanisterControllerButton.spec.ts:91]: Because there is no reset anymore since the "controller" is ephemeral - passed through the event only and not kept in a state

File: frontend/src/lib/types/canister-detail.modal.ts (hunk 3 above)
Comment (peterpeterparker) [reply frontend/src/lib/types/canister-detail.modal.ts:11]: Nice, good idea 👍. No preference so I go for `data` since it's the first you mention but if you rather like `payload` ping me.

File: frontend/src/lib/types/canister-detail.modal.ts (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/lib/types/canister-detail.modal.ts]: Yes, that makes sense to me. It's the usage of the generic type that I found weird. Using `Pick` instead of extending.

File: frontend/src/tests/lib/components/canister-detail/RemoveCanisterControllerButton.spec.ts (hunk 2 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/components/canister-detail/RemoveCanisterControllerButton.spec.ts:91]: Ah, the modal is not in the component and it doesn't open it. Ok.

That probably means that we need to test that the event is emitted then?

Or how do we test that clicking a button opens a specific modal?

**Review by lmuntaner — APPROVED** [review]
Thanks!

File: frontend/src/lib/types/canister-detail.modal.ts (hunk 1 above)
Comment (peterpeterparker) [reply frontend/src/lib/types/canister-detail.modal.ts]: Discussing in chat, modals covered in other spec tests
